### DISCUSSION TITLE
fold and folding in lunarvim

### DISCUSSION THREAD

**Discussion by mbaneshi** [LunarVim/LunarVim#3940] [Q&A] [answered]
But How we can use fold and folding utility in lunarvim ?

**Comment by oscardoso** [comment, marked as answer]
this works for me.

```lua
-- folding powered by treesitter
-- https://github.com/nvim-treesitter/nvim-treesitter#folding
-- look for foldenable: https://github.com/neovim/neovim/blob/master/src/nvim/options.lua
-- Vim cheatsheet, look for folds keys: https://devhints.io/vim
vim.opt.foldmethod = "expr" -- default is "normal"
vim.opt.foldexpr = "nvim_treesitter#foldexpr()" -- default is ""
vim.opt.foldenable = false -- if this option is true and fold method option is other than normal, every time a document is opened everything will be folded.

```

**Comment by aap01** [comment]
Try these settings to fold the methods inside a class when you enter document

```lua
vim.wo.foldmethod = "expr"
vim.wo.foldexpr = "nvim_treesitter#foldexpr()"
vim.wo.fillchars = "fold: "
vim.wo.foldnestmax = 3
vim.wo.foldminlines = 1
vim.wo.foldlevel = 1
vim.wo.foldtext = [[substitute(getline(v:foldstart),'\\t',repeat('\ ',&tabstop),'g').'...'.trim(getline(v:foldend)) . ' (' . (v:foldend - v:foldstart + 1) . ' lines)']]
```

Final outcome

<img width="907" alt="image" src="https://github.com/LunarVim/LunarVim/assets/19292809/37f7457b-2757-4fa4-b59f-6709499596e1">

**Comment by tunix** [comment]
Hi,

Is there a plugin necessary for folding to work? I tried adding the 3 lines inside the accepted answer but I couldn't get it to work. I have a YAML file and I want to fold part of some block. I get my cursor on the first line of the block I want to fold, press z and I only see "spelling" inside the menu.

**Comment by aap01** [comment]
@tunix 

If you try enabling the `z` presets in `whichkey` setup, I think you will be able to see the suggestions for folding and stuff.
Try this in your `config.lua` file
```bash
lvim.builtin.which_key.setup.plugins.presets.z = true
```

**Comment by githorse** [comment]
Note this gotcha: [folding doesn't always work when you initially open the file](https://github.com/nvim-treesitter/nvim-treesitter/issues/1337) -- sometimes you get an error `No folds found`. (Ultimately [caused by Telescope (?)](https://github.com/nvim-telescope/telescope.nvim/issues/699)). 

Simplest workaround is to run `:e` on the file, and then folds will work. Other more permanent workarounds are suggested in the issue.
